synapse/handlers/room_list.py

```diff
@@ -288,23 +292,41 @@ def get_order_for_room(room_id):
 
     @defer.inlineCallbacks
     def _append_room_entry_to_chunk(self, room_id, num_joined_users, chunk, limit,
-                                    search_filter):
+                                    search_filter, from_federation=False):
         """Generate the entry for a room in the public room list and append it
         to the `chunk` if it matches the search filter
         """
         if limit and len(chunk) > limit + 1:
             # We've already got enough, so lets just drop it.
             return
 
-        result = yield self.generate_room_entry(room_id, num_joined_users)
+        if from_federation:
+            result = yield self.generate_room_entry(room_id,
+                self.config.allow_non_federated_in_public_rooms,
+                num_joined_users)
+        else:
+            result = yield self.generate_room_entry(room_id, True, num_joined_users)
```

Comment:
Actually, thinking about this a bit more, I think we can simplify this a lot by having `generate_room_entry` returning the `m.federate` param in the returned `result` and doing the filter here (or in `_matches_room_entry`). That way we don't need to change the cache below